src/hotspot/share/opto/type.cpp

```diff
@@ -1745,218 +1748,120 @@ const TypeInteger* TypeInteger::minus_1(BasicType bt) {
-// Only happens for pessimistic optimizations.
-const Type *TypeInt::narrow( const Type *old ) const {
-  if (_lo >= _hi)  return this;   // already narrow enough
-  if (old == nullptr)  return this;
-  const TypeInt* ot = old->isa_int();
-  if (ot == nullptr)  return this;
-  jint olo = ot->_lo;
-  jint ohi = ot->_hi;
+bool TypeInt::contains(const TypeInt* t) const {
+  return TypeIntHelper::int_type_is_subset(this, t);
+}
 
-  // If new guy is equal to old guy, no narrowing
-  if (_lo == olo && _hi == ohi)  return old;
+const Type* TypeInt::xmeet(const Type* t) const {
+  return TypeIntHelper::int_type_xmeet(this, t, TypeInt::try_make, _is_dual);
+}
 
-  // If old guy was maximum range, allow the narrowing
-  if (olo == min_jint && ohi == max_jint)  return this;
+const Type* TypeInt::xdual() const {
+  return new TypeInt(TypeIntPrototype<jint, juint>{{_lo, _hi}, {_ulo, _uhi}, _bits},
+                     _widen, !_is_dual);
+}
 
-  if (_lo < olo || _hi > ohi)
-    return this;                // doesn't narrow; pretty weird
+const Type* TypeInt::widen(const Type* old, const Type* limit) const {
+  assert(!_is_dual, "dual types should only be used for join calculation");
+  return TypeIntHelper::int_type_widen(this, old->isa_int(), limit->isa_int());
+}
 
-  // The new type narrows the old type, so look for a "death march".
-  // See comments on PhaseTransform::saturate.
-  juint nrange = (juint)_hi - _lo;
-  juint orange = (juint)ohi - olo;
-  if (nrange < max_juint - 1 && nrange > (orange >> 1) + (SMALLINT*2)) {
-    // Use the new type only if the range shrinks a lot.
-    // We do not want the optimizer computing 2^31 point by point.
-    return old;
+const Type* TypeInt::narrow(const Type* old) const {
+  assert(!_is_dual, "dual types should only be used for join calculation");
+  if (old == nullptr) {
+    return this;
   }
 
-  return this;
+  return TypeIntHelper::int_type_narrow(this, old->isa_int());
 }
 
 //-----------------------------filter------------------------------------------
-const Type *TypeInt::filter_helper(const Type *kills, bool include_speculative) const {
+const Type* TypeInt::filter_helper(const Type* kills, bool include_speculative) const {
+  assert(!_is_dual, "dual types should only be used for join calculation");
   const TypeInt* ft = join_helper(kills, include_speculative)->isa_int();
-  if (ft == nullptr || ft->empty())
+  if (ft == nullptr) {
     return Type::TOP;           // Canonical empty value
+  }
+  assert(!ft->_is_dual, "dual types should only be used for join calculation");
   if (ft->_widen < this->_widen) {
     // Do not allow the value of kill->_widen to affect the outcome.
     // The widen bits must be allowed to run freely through the graph.
-    ft = TypeInt::make(ft->_lo, ft->_hi, this->_widen);
+    return (new TypeInt(TypeIntPrototype<jint, juint>{{ft->_lo, ft->_hi}, {ft->_ulo, ft->_uhi}, ft->_bits},
+                        this->_widen, false))->hashcons();
```

Comment:
Why not use `TypeInt::make`?

Reply:
Also: when do you have to do `hashcons` and when is it ok without?

Reply:
> Why not use `TypeInt::make`?

Since we know that the bounds are canonical here, just use the constructor directly saves us the need to do canonicalization.

> Also: when do you have to do `hashcons` and when is it ok without?

We have to use `hashcons` everywhere except at `xdual`.